### PR TITLE
feat(*): implement python client for openGemini

### LINKED ISSUE #1: implement python client

### PR DESCRIPTION
**Pull request by bbddbb1**
support DDL
support write/query
support line protocol
support CI
close #1

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    types: [ opened, synchronize, reopened, ready_for_review ]
+    paths-ignore:
+      - 'config/**'
+      - 'docker/**'
+      - 'docs/**'
+      - 'images/**'
+      - 'scripts/**'
+      - 'tests/**'
+      - '**.md'
+      - '.gitignore'
```
Comment (shilinlee) [inline .github/workflows/ci.yml:15]:
```suggestion
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  push:
+
+jobs:
+  build:
+
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: [ "3.7", "3.8", "3.9", "3.10", "3.11" ]
```
Comment (shilinlee) [inline .github/workflows/ci.yml:24]: Is it necessary to run all of these versions? Is it appropriate to run the minimum version?

**Review by shilinlee — CHANGES_REQUESTED** [review]

File: point.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        if isinstance(key, (list, tuple)) and isinstance(value, (list, tuple)):
+            for k, v in zip(key, value):
+                self._tags[k] = v
+        return self
+
+    def field(self, field, value):
+        """
+        Add field
+        """
+        if isinstance(field, str) and isinstance(value, str):
```
Comment (shilinlee) [inline point.py:107]: field_value should support `string`, `integer`, `bool`, `float`.

File: point.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+            self._fields[field] = value
+        if isinstance(field, (list, tuple)) and isinstance(value, (list, tuple)):
```
Comment (shilinlee) [inline point.py:109]: len(field) should equal len(value)

File: point.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+    if isinstance(timestamp, str):
+        timestamp = parse(timestamp)
+
+    if isinstance(timestamp, datetime):
+        if not timestamp.tzinfo:
+            timestamp = UTC.localize(timestamp)
+
+        ns = _to_nanos(timestamp)
+        if precision is None or precision == 'n':
```
Comment (shilinlee) [inline point.py:54]:
```suggestion
        if precision is None or precision == 'ns':
```

File: point.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+            return ns
+
+        if precision == 'u':
```
Comment (shilinlee) [inline point.py:57]:
```suggestion
        if precision == 's' or precision == 'us' or precision == 'µs':
```

File: point.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+def _convert_timestamp(timestamp, precision=None):
+    if timestamp is None:
+        return ""
+
+    if isinstance(timestamp, int):
+        return timestamp  # assume precision is correct if timestamp is int
+
+    if isinstance(timestamp, str):
+        timestamp = parse(timestamp)
```
Comment (shilinlee) [inline point.py:47]: specify which precision.

File: client.py [hunk 8]
```diff
@@ -0,0 +1,265 @@
+import json
+
+import requests
+
+from point import escape_name
+
+
+class OpenGeminiClient(object):
+    def __init__(
+            self, host, port, username="root", password="root", database=None,
```
Comment (shilinlee) [inline client.py:10]:
```suggestion
            self, host, port, username="", password="", database=None,
```

File: client.py [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+            params_dict.update(bind_params)
+            params["params"] = json.dumps(params_dict)
+        params["q"] = query
+        params["db"] = self._database
+        if (
+                query.lower().startswith("select ") and " into " not in query.lower()
+        ) or query.lower().startswith("show "):
+            method = "GET"
+        else:
+            method = "POST"
```
Comment (shilinlee) [inline client.py:107]: method = "GET"

File: client.py [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+        response = self.request(url="query", params=params, method=method)
+        return response.json()
+
+    def write(self, point):
+        header = self._headers.copy()
+        header["Content-Type"] = "application/octet-stream"
```
Comment (shilinlee) [inline client.py:113]: use `self._write_header`

**Review by shilinlee — CHANGES_REQUESTED** [review]

File: client.py [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+            database = self._database
+        ret = self.query(
+            (f"SHOW RETENTION POLICIES ON "
+             f'"{database.translate(escape_name)}"')
+        )
+        return ret
+
+    def create_user(self, username, password, admin=False):
+        """
+        Create a new user.
```
Comment (shilinlee) [inline client.py:262]:
```suggestion
        Create a new user.
        Create an admin user when admin is True.
        Create a general user when admin is False.
```

File: client.py [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+
+    def ping(self):
+        response = self.request(url="ping", method="GET", expected_response_code=204)
+        return response.headers["X-Geminidb-Version"]
+
+    def create_database(self, dbname):
+        """
+        Create a new database.
+        """
+        self.query(f"CREATE DATABASE {dbname}")
```
Comment (shilinlee) [inline client.py:139]:
```suggestion
        self.query(f"CREATE DATABASE {dbname}", method="POST")
```

File: client.py [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+
+    def drop_database(self, dbname):
+        """
+        Drop a database.
+        """
+        self.query(f"DROP DATABASE {dbname}")
```
Comment (shilinlee) [inline client.py:145]:
```suggestion
        self.query(f"DROP DATABASE {dbname}", method="POST")
```

File: client.py [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+        if duration:
+            query += f" DURATION {duration}"
+        if shard_duration:
+            query += f" SHARD DURATION {shard_duration}"
+        if replication:
+            query += f" REPLICATION {replication}"
+        if default is True:
+            query += " DEFAULT"
+
+        self.query(query)
```
Comment (shilinlee) [inline client.py:229]:
```suggestion
        self.query(query, method="POST")
```

File: client.py [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+        retention policy.
+        :param retention_policy_name: the name of the retention policy to drop
+        :param database: the database for which the retention policy is dropped. Defaults to current client's database
+        """  # noqa: E501
+        if database is None:
+            database = self._database
+        self.query(
+            f'DROP RETENTION POLICY "{retention_policy_name.translate(escape_name)}" '
+            f'ON "{database.translate(escape_name)}"'
+        )
```
Comment (shilinlee) [inline client.py:244]:
```suggestion
method="POST")
        )
```

File: client.py [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+        return ret
+
+    def create_user(self, username, password, admin=False):
+        """
+        Create a new user.
+        """
+        query = f"CREATE USER {username} WITH PASSWORD {password}"
+        if admin:
+            query += " WITH ALL PRIVILEGES"
+        self.query(query)
```
Comment (shilinlee) [inline client.py:267]:
```suggestion
        self.query(query, method="POST")
```

File: client.py [hunk 17]
```diff
@@ -0,0 +1,272 @@
...
+
+    def drop_user(self, username):
+        """Drop a user."""
+        query = f"DROP USER {username}"
+        self.query(query)
```
Comment (shilinlee) [inline client.py:272]:
```suggestion
        self.query(query, method="POST")
```

File: client.py
Comment (shilinlee) [inline client.py:1]: Add your test cases.

**Review by shilinlee — CHANGES_REQUESTED** [review]

File: client.py [hunk 18]
```diff
@@ -0,0 +1,292 @@
...
+            params["params"] = json.dumps(params_dict)
+        params["q"] = query
+        if params.get("db", None) is None:
+            params["db"] = self._database
+        response = self.request(url="query", params=params, method=method)
+        return response.json()
+
+    def write(self, point, precision='ns'):
+        header = self._headers.copy()
+        header["Content-Type"] = "application/octet-stream"
```
Comment (shilinlee) [inline client.py:128]:
```suggestion
```

**Review by shilinlee — CHANGES_REQUESTED** [review]

**Review by shilinlee — APPROVED** [review]

**Review by xiangyu5632 — APPROVED** [review]

**Pull request merged**